File: src/baseliner/engine.py
from __future__ import annotations

import logging
import uuid
from datetime import UTC, datetime

from baseliner.checks.registry import CheckRegistry
from baseliner.models.policy import SEVERITY_WEIGHT, Policy, Severity
from baseliner.models.repository import NormalizedRepository
from baseliner.models.result import CheckResult, CheckStatus, RepoResult, RunResult

LOGGER = logging.getLogger(__name__)


class PolicyEngine:
    def __init__(
        self,
        policy: Policy,
        registry: CheckRegistry,
        global_ignore: list[str] | None = None,
        repo_ignores: dict[str, list[str]] | None = None,
    ) -> None:
        self.policy = policy
        self.registry = registry
        self.global_ignore: set[str] = set(global_ignore or [])
        self.repo_ignores: dict[str, list[str]] = repo_ignores or {}

    def run(self, repo: NormalizedRepository) -> RepoResult:
        repo_ignore_set = set(self.repo_ignores.get(repo.slug, []))
        results: list[CheckResult] = []

        for check_def in self.policy.checks:
            if not check_def.enabled:
                continue
            if check_def.id in self.global_ignore or check_def.id in repo_ignore_set:
                LOGGER.debug("Skipping check '%s' for '%s' (ignored)", check_def.id, repo.slug)
                continue

            try:
                check = self.registry.get(check_def.id)
            except KeyError:
                LOGGER.warning("Unknown check id '%s' in policy — skipping", check_def.id)
                continue

            result = check.evaluate(repo)
            result = result.model_copy(update={"severity": check_def.severity.value})
            results.append(result)

        score = self._compute_score(results)
        return RepoResult(
            slug=repo.slug,
            timestamp=datetime.now(tz=UTC),
            score=score,
            results=results,
        )

    def _compute_score(self, results: list[CheckResult]) -> float:
        total_weight = 0
        passed_weight = 0
        for result in results:
            if result.status == CheckStatus.SKIP:
                continue
            try:
                weight = SEVERITY_WEIGHT[Severity(result.severity)]
            except (ValueError, KeyError):
                weight = 1
            total_weight += weight
            if result.status == CheckStatus.PASS:
                passed_weight += weight
        if total_weight == 0:
            return 1.0
        return round(passed_weight / total_weight, 4)

    def run_batch(self, repos: list[NormalizedRepository]) -> RunResult:
        repo_results: list[RepoResult] = []
        for repo in repos:
            try:
                repo_results.append(self.run(repo))
            except Exception:  # noqa: BLE001
                LOGGER.exception("Unhandled error evaluating repo '%s'", repo.slug)

        passed = sum(
            1
            for repo_result in repo_results
            if all(
                check_result.status not in (CheckStatus.FAIL, CheckStatus.ERROR)
                for check_result in repo_result.results
            )
        )
        failed = len(repo_results) - passed

        return RunResult(
            run_id=str(uuid.uuid4()),
            timestamp=datetime.now(tz=UTC),
            total_repos=len(repo_results),
            passed=passed,
            failed=failed,
            repos=repo_results,
        )

File: internal/engine/engine.go
// Package engine evaluates repositories against a policy and scores them.
package engine

import (
	"crypto/rand"
	"fmt"
	"log/slog"
	"math"
	"time"

	"github.com/baselinerhq/baseliner/internal/checks"
	"github.com/baselinerhq/baseliner/internal/models"
)

// Engine evaluates repos against a policy using a check registry.
type Engine struct {
	Policy       *models.Policy
	Registry     *checks.Registry
	GlobalIgnore map[string]bool
	RepoIgnores  map[string][]string
}

// New builds an engine, normalizing ignore inputs into sets.
func New(policy *models.Policy, registry *checks.Registry, globalIgnore []string, repoIgnores map[string][]string) *Engine {
	gi := make(map[string]bool, len(globalIgnore))
	for _, id := range globalIgnore {
		gi[id] = true
	}
	return &Engine{Policy: policy, Registry: registry, GlobalIgnore: gi, RepoIgnores: repoIgnores}
}

// Run evaluates a single repo and returns its scored result.
func (e *Engine) Run(repo *models.NormalizedRepository, now time.Time) models.RepoResult {
	repoIgnore := make(map[string]bool)
	for _, id := range e.RepoIgnores[repo.Slug] {
		repoIgnore[id] = true
	}

	var results []models.CheckResult
	for _, def := range e.Policy.Checks {
		if !def.Enabled {
			continue
		}
		if e.GlobalIgnore[def.ID] || repoIgnore[def.ID] {
			slog.Debug("skipping ignored check", "check", def.ID, "repo", repo.Slug)
			continue
		}
		c, ok := e.Registry.Get(def.ID)
		if !ok {
			slog.Warn("unknown check id in policy — skipping", "check", def.ID)
			continue
		}
		res := checks.Evaluate(c, repo)
		res.Severity = def.Severity // policy severity overrides the check's default
		results = append(results, res)
	}

	return models.RepoResult{
		Slug:      repo.Slug,
		Timestamp: now,
		Score:     computeScore(results),
		Results:   results,
	}
}

// computeScore is the severity-weighted pass ratio. Skipped checks are
// excluded; an all-skip (zero total weight) result scores 1.0.
func computeScore(results []models.CheckResult) float64 {
	totalWeight, passedWeight := 0, 0
	for _, r := range results {
		if r.Status == models.StatusSkip {
			continue
		}
		w := r.Severity.Weight()
		totalWeight += w
		if r.Status == models.StatusPass {
			passedWeight += w
		}
	}
	if totalWeight == 0 {
		return 1.0
	}
	return math.Round(float64(passedWeight)/float64(totalWeight)*10000) / 10000
}

// RunBatch evaluates many repos and aggregates pass/fail counts.
func (e *Engine) RunBatch(repos []*models.NormalizedRepository, now time.Time) models.RunResult {
	repoResults := make([]models.RepoResult, 0, len(repos))
	for _, repo := range repos {
		repoResults = append(repoResults, e.Run(repo, now))
	}

	passed := 0
	for _, rr := range repoResults {
		if !hasFailureOrError(rr) {
			passed++
		}
	}

	return models.RunResult{
		RunID:      newRunID(),
		Timestamp:  now,
		TotalRepos: len(repoResults),
		Passed:     passed,
		Failed:     len(repoResults) - passed,
		Repos:      repoResults,
	}
}

func hasFailureOrError(rr models.RepoResult) bool {
	for _, r := range rr.Results {
		if r.Status == models.StatusFail || r.Status == models.StatusError {
			return true
		}
	}
	return false
}

// newRunID returns a random UUIDv4 string (avoids an external dependency).
func newRunID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
